lib/net/x86-64-linux/Net.v3

```diff
@@ -0,0 +1,371 @@
+ * for IPv4 should match the struct sockaddr_in type, while the return value
+ * for IPv6 should match the struct sockaddr_in6 type. Note that the port and
+ * IP address need to written in network-byte order, i.e. in big-endian.
+ *
+ * Returns the C-struct representation of {socket}.
+ */
+def socketToSockAddr(socket: Socket, server: bool) -> Array<byte> {
+	match (socket.addr) {
+		Inet4(addrV4) => {
+			// Definition of sockaddr for IPv4:
+			// struct sockaddr_in {
+			//   short int          sin_family;  // Address family, AF_INET
+			//   unsigned short int sin_port;    // Port number
+			//   struct in_addr     sin_addr;    // Internet address
+			//   unsigned char      sin_zero[8]; // Same size as struct sockaddr
+			// };
+			def sockAddrBldr = DataWriter.new();
+			def port = if(server, socket.localPort, socket.port);
+			sockAddrBldr.acquire(LinuxConst.SOCKADDR_V4_SIZE);
+			// sin_family
+			sockAddrBldr.put_b16(LinuxConst.AF_INET);
+			// sin_port
+			sockAddrBldr.put_b16be(int.view(port));
+			// sin_addr
+			for (b in addrV4) {
+				sockAddrBldr.putb(b);
+			}
+			// sin_zero
+			sockAddrBldr.zeroN(8);
+			return sockAddrBldr.extract();
+		}
+		Inet6(addrV6) => {
+			// Definition of sockaddr for IPv6:
+			// struct sockaddr_in6 {
+			//     u_int16_t       sin6_family;   // address family, AF_INET6
+			//     u_int16_t       sin6_port;     // port number, Network Byte Order
+			//     u_int32_t       sin6_flowinfo; // IPv6 flow information (set to 0)
+			//     struct in6_addr sin6_addr;     // IPv6 address
+			//     u_int32_t       sin6_scope_id; // Scope ID (set to 0)
+			// };
+			//
+			// struct in6_addr {
+			//     unsigned char   s6_addr[16];   // IPv6 address
+			// };
+			def sockAddrBldr = DataWriter.new();
+			def port = if(server, socket.localPort, socket.port);
+			sockAddrBldr.acquire(LinuxConst.SOCKADDR_V6_SIZE);
+			// sin6_family
+			sockAddrBldr.put_b16(LinuxConst.AF_INET6);
+			// sin6_port
+			sockAddrBldr.put_b16be(int.view(port));
+			// sin6_flowinfo
+			sockAddrBldr.zeroN(4);
+			// sin6_addr
+			for (b in addrV6) {
+				sockAddrBldr.putb(b);
+			}
+			// sin6_scope_id
+			sockAddrBldr.zeroN(4);
+			return sockAddrBldr.extract();
+		}
+	}
+}
+
+/**
+ * An abstraction over a network socket that allows for sending and receiving
+ * messages.
+ */
+class Socket {
+	// IP address of socket
+	def addr: InetAddress;
+	// Associated file descriptor of socket
+	def var fd: int;
+	// Local port the socket is connected to
+	def var localPort: int;
+	// Server port the socket is connected to
+	def var port: int;
+
+	new(addr, localPort, port, sockfd: int, createSocket: bool) {
+		if (createSocket) {
+			def family = if(InetAddress.Inet4.?(addr), LinuxConst.AF_INET, LinuxConst.AF_INET6);
+			fd = Net.socket(family, LinuxConst.SOCK_STREAM /* type */, 0 /* protocol */);
+		} else {
+			fd = sockfd;
+		}
+	}
+
+	def setLocalPort(local: int) {
+		localPort = local;
+	}
+
+	/**
```

Comment:
Can you use `//` comments? I tend to prefer saving lines.

Reply:
Done.